### ISSUE TITLE
Android 5.0 Lollipop / ART support

### ISSUE THREAD

**Issue by celestefox** [dyne/ZShaolin#6]
The current version of Zshaolin in the Google Play store refuses to be installed on Android Lollipop, saying my device isn't compatible with this version. While the latest changelog notes Android Lollipop isn't supported, I thought I'd open this issue anyways (for tracking progress/just so it's clear this is known).

**Comment by jaromil** [comment]
Yep. Incompatibility is due to PIE executable restriction in Lollypop, which I guess rules out our current toolchain.

FYI users reported ZShaolin to work on Cyanogen Mod based on Lollypop.

I'm investigating the issue, but cannot commit to have a fix for it anytime soon.

**Comment by timoc** [comment]
+1
only some executables are broken. zsh for example works. presumably it can be mitigated with static linking?

**Comment by jaromil** [comment]
ZSh did not work in my tests on unroofed Android 5.

Can you provide details on your testing environment? version, root and anything else relevant.... thanks

**Comment by timoc** [comment]
it was an upgrade from 4.4 rather than an install, so unsure if that makes a difference. 
Sony Xperia Z Ultra 5.0.2 
3.0.4-perf
Not sure what else i can give you.

**Comment by jaromil** [comment]
is it rooted?
static linking is not a solution: ZShaolin is already all statically linked with our own toolchain.
The fix for Android 5 includes rebuilding everything with -fPIE.
Unfortunately I have no time to fix it now, hopefully soon.

**Comment by timoc** [comment]
Nope not rooted
Have not tried building anything myself.

**Comment by jesshbrewer** [comment]
Please let us know if you get it working on 5.1
It's frustrating to have the golden apple dangling just out of reach....

**Comment by timoc** [comment]
I've not done anything new. any news on a 5.x version?

**Comment by ghost** [comment]
+1

Tried to compile, but some packages are missing from the package folder on code.dyne.org. (404)

Any directions on how to do this?

**Comment by jaromil** [comment]
list of missing packages?

I have very little time this month, but a rebuild for lollypop is in my todo

any hints or help appreciated for now, but I cannot provide help myself (too many other work commitments!)

**Comment by ghost** [comment]
Thank you for your reply.

Here is a list of some problems (a lot of problems):
- `zmake build all` does not work, but it does if you replace `all` with the name of a folder in `build/`
  - On a side note, I think this is better because you do not need to build everything
- Missing packages (the REPO is https://releases.dyne.org/zshaolin/sources/)
- General compiling errors

**CRYPTO** module
- Missing packages:
  - gmp-5.1.3.tar.lz
  - nettle-2.7.1.tar.gz
  - libidn-1.28.tar.gz
  - gnutls-3.1.16.tar.xz
  - mcrypt-2.6.8.tar.gz
  - srm-1.2.14.tar.gz

**GNOMISH** module
- _glib-2.39.1.tar.xz_ `configure: error: cannot run test program while cross compiling`
  - `config.log` available for debugging if you want
- _mc-4.8.10.tar.bz2_ `configure: error: could not determine how to read list of mounted file systems`
  - `config.log` available for debugging if you want

**LANG** module
- No problems :+1: 

**MEDIA** module
- Missing packages:
  - jhead-2.97.tar.gz
- Compiling problem: _libpng_: `zlib not installed`
  - Needs some custom `--libdir` flag, no clue where to insert it, I'm using Arch Linux x64
  - This seems to trigger a cascading build failure

**NETWORK** module
- Can't find the Android NDK toolchain, resolved reextracting the NDK toolchain to `toolchains/`
- Missing packages:
  - lynx-2.8.7.tar.gz
  - ncdu-1.10.tar.gz
  - libuuid-1.0.3.tar.gz
  - testdisk-6.14.tar.bz2
- _curl_ `configure: error: C compiler cannot create executables`
  - `config.log` available for debugging if you want
- _git_ Same error as above
  - `config.log` available for debugging if you want
- _ncurses_: `../form/fty_num.c:226:14: error: 'struct lconv' has no member named 'decimal_point'`
  - Error with the library `fty_num.o`, make failed

**SOUND** module
- _alsa-utils_: `panel.h not found`, `error: required curses helper header not found`
  - Seems it can't find a ncurses header, but I have it in `/usr/include/`
  - Will try symlinking as per here: https://bugs.archlinux.org/task/8853

**SYSTEM** module
- Missing packages:
  - pcre-8.35.tar.bz2
  - zsh-5.0.6.tar.bz2
  - tmux-1.9a.tar.gz
- Cascading build failure (again)

**TEXT** module
- Missing packages:
  - vim-7.3.tar.bz2
  - discount-git.tar.gz
- _mdocml_: `configure not found`

**TOYS** module :balloon: 
- Missing packages:
  - curseofwar-1.2.tar.gz
  - ctris-0.42.tar.bz2
  - cryptoslam-1.2.tar.gz
- _libcaca_: `error: cannot find slang development files`
  - Same problem as _alsa-utils_
- _matanza_: `ctangle matanza.w: command not found`
  - On my system, `ctangle` is included in the `texlive-bin` package
  - This is not actually a problem

Sorry for the lengthy comment, but I wanted to be as explicative as possible.
If you have any suggestions, let me know.

**Comment by jaromil** [comment]
Thanks @pellettiero this is really helpful and will save me a lot of time going through the missing packages. I am still swamped right now, but the task is way more approachable following your notes.

Since release version 1.0 I had to dismiss the old machine used to build ZShaolin (OS was ubuntu 12.04) but still have its harddisk and will at least upload missing packages from it.

Regarding the errors yes, I remember having fixed those by hand with very small and annoying changes to configure files which at the end were difficult to script into the zmake. Most of them are due to cross-compiling and test files that cannot be run on build host...

**Comment by ghost** [comment]
No problem @jaromil thanks to you.
Also, did you check out a similar project that seems to be going pretty well and maybe that'll give you some ideas?
http://termux.com/
https://github.com/termux

Hope you'll resolve those issues! If you need help, I'm here!

**Comment by jaromil** [comment]
thanks for pointing to termux. does that covers all functionalities in zshaolin?
if yes I'd rather recommend using that and stop developing zshaolin

still proud to have pioneered the field,as it was 15 years ago for livecd tech ;-)

**Comment by jaromil** [comment]
tried termux and must say it has some rough edges, keyb popup is stuck sometimes....

but thr fact it has apt repo makes it look like it can be a long standing implementation

I'll keep an eye and see. Perhaps zshaolin can use termux repos, but so far its keyboard and the ad-hoc configuration of programs makes it worthed to have still. Termux seems like a generic distro installation approach without much artisanal work on it.

**Comment by fornwall** [comment]
Hi, author of Termux here :)!

> tried termux and must say it has some rough edges, keyb popup is stuck sometimes....

The keyboard can be toggled with the keyboard button on the left slideout menu. I just released version 0.11 to the play store (available in a couple of hours) which adds toggling the keyboard visibility by tapping on the terminal!

I'm happy to hear additional feedback in how Termux could be improved! Do you have any tips and program configurations from zshaolin that can be brought over to Termux?

**Comment by ghost** [comment]
Yeah, you're right @jaromil.
An APT implementation would be great!

**Comment by jaromil** [comment]
Hi @fornwall ! ATM I'm busy on other projects, but as soon as I'll be busy on ZShaolin will certainly update you on ideas and configurations, after going through more testing of termux to understand what we can share.
